# Завдання 7

In [2]:
import random

def simulate_dice_rolls(num_rolls):
    sums_count = {i: 0 for i in range(2, 13)}  # словник для підрахунку кількості кожної суми

    for _ in range(num_rolls):
        dice1 = random.randint(1, 6)  # кидок першого кубика
        dice2 = random.randint(1, 6)  # кидок другого кубика
        total = dice1 + dice2
        sums_count[total] += 1  # оновлення кількості випадінь суми

    probabilities = {k: v / num_rolls * 100 for k, v in sums_count.items()}  # обчислення ймовірностей у відсотках
    return probabilities

# Виконати симуляцію
num_rolls = 1000000  # велика кількість кидків кубиків
simulated_probabilities = simulate_dice_rolls(num_rolls)

# Вивести результати симуляції
print("Сума\tІмовірність")
for total, probability in simulated_probabilities.items():
    print(f"{total}\t{probability:.2f}% ({simulated_probabilities[total]}/{num_rolls})")


Сума	Імовірність
2	2.79% (2.7869/1000000)
3	5.56% (5.5564/1000000)
4	8.34% (8.3425/1000000)
5	11.13% (11.1297/1000000)
6	13.87% (13.866800000000001/1000000)
7	16.64% (16.6438/1000000)
8	13.90% (13.8991/1000000)
9	11.12% (11.1167/1000000)
10	8.31% (8.3074/1000000)
11	5.57% (5.5653/1000000)
12	2.79% (2.7854/1000000)


Після тестування методом Monte Carlo можна порівняти результати з аналітичним підрахунком ймовірности

| Сума | Теоретична Імовірність (n/загальна кількість варіантів) | Імовірність методом Monte Carlo (значення/кількість тестувань) |
|------|---------------------------------------------------------|----------------------------------------------------------------|
| 2    | 2.78% (1/36)                                            | 2.79% (2.7869/1000000)                                         |
| 3    | 5.56% (2/36)                                            | 5.56% (5.5564/1000000)                                         |
| 4    | 8.33% (3/36)                                            | 8.34% (8.3425/1000000)                                         |
| 5    |11.11% (4/36)                                            | 11.13% (11.1297/1000000)                                       |
| 6    |13.89% (5/36)                                            | 13.87% (13.8668/1000000)                                       |
| 7    |16.67% (6/36)                                            | 16.64% (16.6438/1000000)                                       |
| 8    |13.89% (5/36)                                            | 13.90% (13.8991/1000000)                                       |
| 9    |11.11% (4/36)                                            | 11.12% (11.1167/1000000)                                       |
|10    | 8.33% (3/36)                                            | 8.31% (8.3074/1000000)                                         |
|11    | 5.56% (2/36)                                            | 5.57% (5.5653/1000000)                                         |
|12    | 2.78% (1/36)                                            | 2.79% (2.7854/1000000)                                         |

# Висновок
Як видно з таблиці, значення отримані за допомогою методу Monte Carlo майже не відрізняються від очікуваних значень отриманих аналітично. Максимальна розбіжність це 3 сотих і тільки для одного значення. Результати показали що метод Monte Carlo можна використовувати для обчислення ймовірності настання певної події і він дає досить точний результат при умові що у нас достатня кількість експерементів
